lib/framedivider.rb

```diff
@@ -390,46 +422,284 @@ def opening(model, id)
+  subsurfaces.each do |s|
+    id = s.nameString
+
+    unless s.vertices.size == 3 || s.vertices.size == 4
+      msg = "Subsurface '#{id}' vertex count (must be 3 or 4) - skipping"
+      TBD.log(TBD::ERROR, msg)
+      next
+    end
+
+    typ = s.subSurfaceType.downcase
+    type = :skylight
+    type = :window if typ.include?("window")
+    type = :door if typ.include?("door")
+    glazed = true if type == :door && typ.include?("glass")
+
+    gross = s.grossArea
+    if gross < TOL
+      msg = "Subsurface '#{id}' gross area (< TOL) - skipping"
+      TBD.log(TBD::ERROR, msg)
+      next
+    end
+
+    c = s.construction
+    if c.empty?
+      msg = "Subsurface '#{id}' missing construction - skipping"
+      TBD.log(TBD::ERROR, msg)
+      next
+    end
+    c = c.get.to_Construction.get
+
+    # A subsurface may have an overall U-factor set by the user - a less
+    # accurate option, yet easier to process (and often the only option
+    # available). With EnergyPlus' "simple window" model, a subsurface's
+    # construction has a single SimpleGlazing material/layer holding the whole
+    # product U-factor.
+    #
+    #   https://bigladdersoftware.com/epx/docs/9-5/engineering-reference/
+    #   window-calculation-module.html#simple-window-model
+    #
+    # In other cases, TBD will recover an 'additional property' tagged
+    # "uFactor", assigned either to the individual subsurface itself, or else
+    # assigned to its referenced construction (a more generic fallback).
+    #
+    # If all else fails, TBD will calculate an approximate whole product
+    # U-factor by adding up the subsurface's construction material thermal
+    # resistances (as well as the subsurface's parent surface film resistances).
+    # This is the least accurate option, especially if subsurfaces have Frame
+    # & Divider objects.
+
+    u = s.uFactor
+    u = s.additionalProperties.getFeatureAsDouble("uFactor") if u.empty?
+    u = c.additionalProperties.getFeatureAsDouble("uFactor") if u.empty?
+    if u.empty?
+      r = rsi(c, surface.filmResistance)
+      if r < TOL
+        msg = "Subsurface '#{id}' U-factor unavailable - skipping"
+        TBD.log(TBD::ERROR, msg)
+        next
+      else
+        u = 1.0 / r
+      end
+    else
+      u = u.get
+    end
+
+    # Should verify convexity of vertex wire/face ...
+    #
+    #       A
+    #      / \
+    #     /   \
+    #    /     \
+    #   / C --- D    <<< allowed as OpenStudio/E+ subsurface?
+    #  / /
+    #  B
+    #
+    # Should convert (annoying) 4-point subsurface into triangle ...
+    #        A
+    #       / \
+    #      /   \
+    #     /     \
+    #    B - C - D   <<< allowed as OpenStudio/E+ subsurface?
+    #
+    four = true if s.vertices.size == 4
+
+    if s.windowPropertyFrameAndDivider.empty?
+      vec = s.vertices
+      area = gross
+    else
+      width = s.windowPropertyFrameAndDivider.get.frameWidth
+      ptz = offset(s.vertices, width)
```

Comment:
I'm not sure if you would want to add `ptz` as a new wire in the Topolys Model?  It should intersect or update other wires  automatically, I guess it depends if you want to do more with it than get the area

Reply:
If the Frame & Divider inputs are valid for a given surface, then those are the points that Topolys gets as input (not the original OpenStudio subsurface vertices). If there's a conflict anywhere, then Topolys gets the original vertices.